ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Tables/Diff.hs

```diff
@@ -0,0 +1,227 @@
+                    else Just (Insert v2))
+      m1
+      m2
+
+fromMap :: Map k (Delta v) -> Diff k v
+fromMap = Diff
+
+fromMapInserts :: Map k v -> Diff k v
+fromMapInserts = Diff . Map.map Insert
+
+fromMapDeletes :: Map k v -> Diff k v
+fromMapDeletes = Diff . Map.map (const Delete)
+
+fromSetDeletes :: Set k -> Diff k v
+fromSetDeletes = Diff . Map.fromSet (const Delete)
+
+fromList :: Ord k => [(k, Delta v)] -> Diff k v
+fromList = Diff . Map.fromList
+
+fromListInserts :: Ord k => [(k, v)] -> Diff k v
+fromListInserts = Diff . Map.fromList . fmap (second Insert)
+
+fromListDeletes :: Ord k => [(k, v)] -> Diff k v
+fromListDeletes = Diff . Map.fromList . fmap (second (const Delete))
+
+{------------------------------------------------------------------------------
+  Query
+------------------------------------------------------------------------------}
+
+null :: Diff k v -> Bool
+null (Diff m) = Map.null m
+
+size :: Diff k v -> Int
+size (Diff m) = Map.size m
+numInserts :: Diff k v -> Int
+numInserts (Diff m) = getSum $ foldMap' f m
+  where
+    f (Insert _) = 1
+    f Delete     = 0
+
+numDeletes :: Diff k v -> Int
+numDeletes (Diff m) = getSum $ foldMap' f m
+  where
+    f (Insert _) = 0
+    f Delete     = 1
+
+{------------------------------------------------------------------------------
+  Applying diffs
+------------------------------------------------------------------------------}
+
+applyDiff ::
+     Ord k
+  => Map k v
+  -> Diff k v
+  -> Map k v
+applyDiff m (Diff diffs) =
+    Merge.merge
+      Merge.preserveMissing
+      (Merge.mapMaybeMissing newKeys)
+      (Merge.zipWithMaybeMatched oldKeys)
+      m
+      diffs
+  where
+    newKeys :: k -> Delta v -> Maybe v
+    newKeys _k (Insert x) = Just x
+    newKeys _k Delete     = Nothing
+
+    oldKeys :: k -> v -> Delta v -> Maybe v
+    oldKeys _k _v1 (Insert x) = Just x
+    oldKeys _k _v1 Delete     = Nothing
+
+applyDiffForKeys ::
+     Ord k
+  => Map k v
+  -> Set k
+  -> Diff k v
+  -> Map k v
+applyDiffForKeys m ks (Diff diffs) =
+  applyDiff
+    m
+    (Diff $ diffs `Map.restrictKeys` (Map.keysSet m `Set.union` ks))
+
+{-------------------------------------------------------------------------------
+  Filter
+-------------------------------------------------------------------------------}
+
+filterOnlyKey :: (k -> Bool) -> Diff k v -> Diff k v
+filterOnlyKey f (Diff m) = Diff $ Map.filterWithKey (const . f) m
+
+{-------------------------------------------------------------------------------
+  From-to anti-diffs
+-------------------------------------------------------------------------------}
+
+fromAntiDiff :: Anti.Diff k v -> Diff k v
+fromAntiDiff (Anti.Diff d) = Diff (Map.map (f . Anti.last) d)
+  where
+    f (Anti.Insert v) = Insert v
+    f Anti.Delete{}   = Delete
+
+toAntiDiff :: Diff k v -> Anti.Diff k v
+toAntiDiff (Diff d) = Anti.Diff (Map.map f d)
+  where
+    f (Insert v) = Anti.singletonInsert v
+    f Delete     = Anti.singletonDelete
```

Comment:
Related to my previous comment, but we can probably move these into `DiffSeq` as well, because these functions are only used in conjunction with a `DiffSeq`